cloudferrylib/utils/retrying.py

```diff
@@ -0,0 +1,197 @@
+                 predicate=None,
+                 predicate_args=None,
+                 predicate_kwargs=None,
+                 predicate_retval_as_arg=False,
+                 retry_message=None,
+                 stop_on_return_value=False):
+        self.reraise_original_exception = reraise_original_exception
+        self.timeout = timeout
+        self.total_time = 0
+        self.attempt = 0
+        self.max_attempts = max_attempts
+        self.wait_exponential = wait_exponential
+        self.max_time = max_time
+        self.raise_error = raise_error
+        self.retry_on_return_value = retry_on_return_value
+        self.return_value = return_value
+        self.expected_exceptions = expected_exceptions or []
+        self.predicate = predicate
+        self.predicate_args = predicate_args
+        self.predicate_kwargs = predicate_kwargs
+        self.predicate_retval_as_arg = predicate_retval_as_arg
+        self.retry_message = retry_message
+        self.stop_on_return_value = stop_on_return_value
+
+    def max_attempts_reached(self):
+        return self.attempt >= self.max_attempts
+
+    def increment_attempts(self):
+        self.attempt += 1
+
+    def timedout(self):
+        return self.total_time >= self.max_time
+
+    def update_total_time(self):
+        if self.wait_exponential:
+            self.timeout *= 2
+
+        self.total_time += self.timeout
+
+    def sleep(self):
+        time.sleep(self.timeout)
+
+    def get_retry_message(self, auto_message):
+        if self.retry_message:
+            return self.retry_message
+        else:
+            return auto_message
+
+    def handle_predicate(self, retval):
+        if self.predicate and callable(self.predicate):
+            if self.predicate_retval_as_arg:
+                args = [retval]
+            else:
+                args = self.predicate_args or []
+            kwargs = self.predicate_kwargs or {}
+            if not self.predicate(*args, **kwargs):
+                default_msg = "Function exited with '{}'".format(retval)
+                msg = self.get_retry_message(default_msg)
+                raise RetryRequired(msg)
+
+    def handle_return_on_value(self, retval):
+        if self.stop_on_return_value and retval != self.return_value:
+            msg = "Return value '{}' didn't match expected '{}'"
+            raise RetryRequired(msg.format(retval, self.return_value))
+
+        if self.retry_on_return_value and retval == self.return_value:
+            msg = self.get_retry_message(
+                "Unexpected return value: {}".format(retval))
+            raise RetryRequired(msg)
+
+    def run(self, func, *args, **kwargs):
+        stop_retrying = self.max_attempts_reached
+        step_action = self.increment_attempts
+
+        if self.max_time is not UNLIMITED:
+            stop_retrying = self.timedout
+            step_action = self.update_total_time
+
+        retval = None
+        failing = True
+        last_exception = None
```

Comment:
we need to have an information about first exception. As an example: port in use is last exception, but we need to know, what happen at first time.